File: examples/enterprise/python/ent_03_cdc_debezium_to_delta.py
"""Enterprise 03 · CDC (Debezium/Kafka) → Delta Lake

Reads Debezium change events from the "pgserver.public.orders" Kafka topic,
parses Insert/Update/Delete operations, and merges them into a Delta table
using the deltalake Python library.

Schema evolution: the orders table gains a "notes" column mid-stream — the
Delta merge handles nullable new columns automatically.

Prerequisites:
    make infra-up debezium-up
    # INSERT/UPDATE some rows in PostgreSQL to generate CDC events:
    # docker exec -it krishiv-postgres psql -U krishiv enterprise -c \
    #   "INSERT INTO orders(customer,product,amount,status) VALUES('alice','Laptop',1299,'pending');"

Run:
    KAFKA_BOOTSTRAP=localhost:9092 python python/ent_03_cdc_debezium_to_delta.py
"""

import json
import os
import time
from typing import Any

import pyarrow as pa
import pyarrow.parquet as pq
from confluent_kafka import Consumer, KafkaException
from deltalake import DeltaTable, write_deltalake

BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP", "localhost:9092")
TOPIC = "pgserver.public.orders"
GROUP_ID = "krishiv-ent-py-03"
DELTA_PATH = "/tmp/krishiv-enterprise-py-03-cdc-delta"
TIMEOUT_S = 20


# Synthetic events matching what Debezium 2.x produces on pgserver.public.orders.
SYNTHETIC_EVENTS = [
    {"op": "c", "order_id": 1, "customer": "alice",  "product": "Laptop Pro", "amount": 1299.99, "status": "pending", "__lsn": 100, "__ts_ms": 1716201600000},
    {"op": "c", "order_id": 2, "customer": "bob",    "product": "Mouse",      "amount":   29.99, "status": "pending", "__lsn": 110, "__ts_ms": 1716201601000},
    {"op": "u", "order_id": 1, "customer": "alice",  "product": "Laptop Pro", "amount": 1299.99, "status": "shipped", "__lsn": 120, "__ts_ms": 1716201602000},
    {"op": "c", "order_id": 3, "customer": "carol",  "product": "Monitor",    "amount":  499.99, "status": "pending", "notes": "gift", "__lsn": 130, "__ts_ms": 1716201603000},
    {"op": "d", "order_id": 2, "customer": None,     "product": None,         "amount": None,    "status": None,      "__lsn": 140, "__ts_ms": 1716201604000},
]


def main() -> None:
    print("=== Enterprise 03 (Python): CDC (Debezium) → Delta Lake ===")
    print(f"  source : kafka://{BOOTSTRAP}  topic={TOPIC}")
    print(f"  sink   : {DELTA_PATH}")

    events = _collect_kafka_events()
    if not events:
        print("  no live CDC events — using synthetic Debezium payloads")
        events = SYNTHETIC_EVENTS

    inserts = [e for e in events if e["op"] == "c"]
    updates = [e for e in events if e["op"] == "u"]
    deletes = [e for e in events if e["op"] == "d"]
    print(f"\n  events: {len(inserts)} inserts, {len(updates)} updates, {len(deletes)} deletes")

    # Write initial snapshot (inserts + updates) to Delta.
    all_rows = [e for e in events if e["op"] in ("c", "u")]
    if all_rows:
        table = _events_to_arrow(all_rows)
        print(f"\n  writing {table.num_rows} rows to Delta (mode=overwrite for demo)…")
        write_deltalake(DELTA_PATH, table, mode="overwrite")

    # Append a delete-marker batch (includes _op column).
    if deletes:
        del_table = _events_to_arrow(deletes)
        write_deltalake(DELTA_PATH, del_table, mode="append")

    # Read back.
    dt = DeltaTable(DELTA_PATH)
    result = dt.to_pyarrow_table()
    print(f"\n--- Delta table ({result.num_rows} rows, {dt.version()} versions) ---")
    import pandas as pd
    df = result.to_pandas()
    print(df.to_string(index=False))

    # Operation summary.
    print("\n--- CDC operation distribution ---")
    print(df.groupby("op")[["order_id"]].count().rename(columns={"order_id": "count"}))

    print(f"\n✓ Delta table at {DELTA_PATH}  (version {dt.version()})")


def _collect_kafka_events(timeout_s: float = 10.0) -> list[dict]:
    consumer = Consumer({
        "bootstrap.servers": BOOTSTRAP,
        "group.id": GROUP_ID,
        "enable.auto.commit": True,
        "auto.offset.reset": "earliest",
    })
    consumer.subscribe([TOPIC])
    events: list[dict] = []
    deadline = time.monotonic() + timeout_s
    try:
        while time.monotonic() < deadline:
            msg = consumer.poll(timeout=1.0)
            if msg is None:
                continue
            if msg.error():
                break
            try:
                payload = json.loads(msg.value().decode("utf-8"))
                payload["__lsn"] = payload.get("__lsn") or 0
                payload["__ts_ms"] = payload.get("__ts_ms") or 0
                payload["op"] = payload.get("__op") or payload.get("op", "c")
                events.append(payload)
                print(f"  kafka event  op={payload['op']}  offset={msg.offset()}")
            except Exception:
                continue
    finally:
        consumer.close()
    return events


def _events_to_arrow(events: list[dict]) -> pa.Table:
    # Collect union of all keys.
    all_keys = sorted({k for e in events for k in e})
    cols: dict[str, list[Any]] = {k: [] for k in all_keys}
    for e in events:
        for k in all_keys:
            cols[k].append(e.get(k))
    return pa.table(cols)


if __name__ == "__main__":
    main()

File: examples/enterprise/rust/src/bin/ent_03_cdc_debezium_to_delta.rs
//! Enterprise 03 · CDC (Debezium) → Delta Lake — embedded mode
//!
//! Uses `InMemoryCdcEventSource` with synthetic Debezium 2.x JSON payloads.
//! The pipeline parses Insert/Update/Delete events, builds an Arrow batch
//! (with `_op`, `_lsn`, `_ts_ms` metadata columns), and writes it to a
//! local Delta table via `write_delta`.
//!
//! In production, swap `InMemoryCdcEventSource` for `RdkafkaCdcEventSource`
//! pointed at your Debezium Kafka topic.
//!
//! Run:
//!   cargo run -p krishiv-enterprise-examples --bin ent_03_cdc_debezium_to_delta

use anyhow::{Context, Result};
use krishiv_connectors::cdc::{
    CdcToLakehousePipeline, InMemoryCdcEventSource, build_batch_from_events,
    parse_debezium_envelope,
};
use krishiv_connectors::lakehouse::{DeltaWriteMode, write_delta};
use tempfile::tempdir;

const EVENTS: &[&str] = &[
    r#"{"op":"c","source":{"lsn":100,"ts_ms":1716201600000,"table":"orders"},"after":{"order_id":1,"customer":"alice","product":"Laptop Pro","amount":"1299.99","status":"pending"}}"#,
    r#"{"op":"c","source":{"lsn":110,"ts_ms":1716201601000,"table":"orders"},"after":{"order_id":2,"customer":"bob","product":"Mouse","amount":"29.99","status":"pending"}}"#,
    r#"{"op":"u","source":{"lsn":120,"ts_ms":1716201602000,"table":"orders"},"before":{"order_id":1},"after":{"order_id":1,"customer":"alice","product":"Laptop Pro","amount":"1299.99","status":"shipped"}}"#,
    // schema evolution: "notes" column appears on order 3
    r#"{"op":"c","source":{"lsn":130,"ts_ms":1716201603000,"table":"orders"},"after":{"order_id":3,"customer":"carol","product":"Monitor 4K","amount":"499.99","status":"pending","notes":"gift"}}"#,
    r#"{"op":"d","source":{"lsn":140,"ts_ms":1716201604000,"table":"orders"},"before":{"order_id":2},"after":null}"#,
];

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Enterprise 03: CDC (Debezium) → Delta Lake (embedded) ===");

    let dir = tempdir()?;
    let delta_path = dir.path().join("orders_delta");
    println!("  delta path : {}", delta_path.display());

    // Parse all events.
    let mut events = Vec::new();
    for (i, json) in EVENTS.iter().enumerate() {
        let event = parse_debezium_envelope(json, 0, i as i64)
            .map_err(|e| anyhow::anyhow!("parse error on event {}: {}", i, e))?;
        println!("  event {}  op={:?}  table={}", i, event.op, event.table);
        events.push(event);
    }

    // Build single Arrow batch from all CDC events.
    let batch = build_batch_from_events(&events)
        .map_err(|e| anyhow::anyhow!("build_batch_from_events: {}", e.0))?;

    println!(
        "\n  merged batch: {} rows  columns: {:?}",
        batch.num_rows(),
        batch.schema().fields().iter().map(|f| f.name().as_str()).collect::<Vec<_>>()
    );

    // Write to Delta Lake.
    write_delta(
        delta_path.to_string_lossy().to_string(),
        vec![batch],
        DeltaWriteMode::Append,
        false,
    )
    .await
    .context("write_delta")?;

    println!("✓ Delta table written");

    // Read back and show results.
    let session = krishiv::Session::builder().build()?;
    let df = session.read_delta_async(delta_path.to_string_lossy().as_ref(), None).await?;
    println!("\n--- CDC output (all ops) ---");
    println!("{}", df.collect()?.pretty()?);

    // Operation distribution.
    let session2 = krishiv::Session::builder().build()?;
    let df2 = session2.read_delta_async(delta_path.to_string_lossy().as_ref(), None).await?;
    let result = df2.collect()?;
    session2.register_record_batches("cdc_out", result.into_batches())?;
    let summary = session2.sql("SELECT _op, COUNT(*) AS n FROM cdc_out GROUP BY _op ORDER BY n DESC")?;
    println!("\n--- Op summary ---");
    println!("{}", summary.collect()?.pretty()?);

    // Also demonstrate the pipeline approach with InMemoryCdcEventSource.
    println!("\n--- InMemoryCdcEventSource pipeline ---");
    let source = InMemoryCdcEventSource::new(EVENTS.iter().copied());
    let pipeline = CdcToLakehousePipeline::new(
        "orders.cdc",
        vec!["broker:9092".into()],
        "my_catalog",
        "warehouse.orders",
        vec!["order_id".into()],
    );
    pipeline.validate().context("pipeline validate")?;
    println!("  pipeline config valid ✓");

    let (shutdown_tx, shutdown_rx) = tokio::sync::watch::channel(false);
    let mut pipeline_batches = Vec::new();
    pipeline
        .run_with_source(
            source,
            |b| { pipeline_batches.push(b); Ok(()) },
            shutdown_rx,
        )
        .await
        .context("run_with_source")?;
    drop(shutdown_tx);

    let total: usize = pipeline_batches.iter().map(|b| b.num_rows()).sum();
    println!("  pipeline produced {} rows across {} batches", total, pipeline_batches.len());

    Ok(())
}
